### DISCUSSION TITLE
Adding `priority` prop for `<Script />` component ⚡

### DISCUSSION THREAD

**Discussion by sanjaiyan-dev** [vercel/next.js#39384] [Ideas]
## Adding `priority` prop for `<Script />` component for better loading performance (Inspired from `next/image`).

Hi, we can `preload` or `prefetch` the scripts by adding `priority` prop which helps to enhance the loading performance of the third-party scripts. 
We can `prefetch` the script if  `strategy` is `lazyOnLoad` or else we can `preload` the scripts.

We can `preload` the scripts which have strategy of `beforeInteractive` or `afterInteractive` since Google Chrome advices that preloaded scripts should be used before 3s. And we can `prefetch` the scripts with strategy of `lazyOnLoad` hence this might take longer than 3s to append the script. 

### Caveat & Solutions
This have significant network bandwidth issue if it is not used wisely and we prefer it to use it with `beforeInteractive` strategy (only when 100% necessary) and we can mark it as `fetchpriority` (new HTML related attribute) `low` to not to collide with hydration related scripts. 

### Need suggestions
1) Do we `prefetch` even the scripts with strategy of `afterInteractive` ?
2) Any other ideas ?

**Comment by sanjaiyan-dev** [comment]
**Related PR -: https://github.com/vercel/next.js/pull/39373**

**Comment by ijjk** [comment]
I don't think we need a separate `priority` prop for `next/script` as I think we should only prefetch where we see an overall performance benefit with specific loading strategies cc @housseindjirdeh for thoughts

**Comment by housseindjirdeh** [comment]
Agreed with @ijjk. We can consider using `prefetch`, `preload`, and/or `fetchpriority` heuristics for the different script loading strategies but I don't think we should have a separate `priority` prop for this. This will only add an extra layer of developer input required which really isn't necessary.

@sanjaiyan-dev Have you put together any tests that shows a performance win when either of these resource hints or properties are used (via Lighthouse, WebPageTest, Chrome DevTools, etc.)? It would be good to get some data first to validate :)

**Reply by sanjaiyan-dev** [reply]
> Agreed with @ijjk. We can consider using `prefetch`, `preload`, and/or `fetchpriority` heuristics for the different script loading strategies but I don't think we should have a separate `priority` prop for this. This will only add an extra layer of developer input required which really isn't necessary.
> 
> @sanjaiyan-dev Have you put together any tests that shows a performance win when either of these resource hints or properties are used (via Lighthouse, WebPageTest, Chrome DevTools, etc.)? It would be good to get some data first to validate :)

Hi this will act similar to priority prop on `<Image />`, we can preload hight priority scripts for faster load time.


_Sorry if I made any mistake in my above comments :(_

**Comment by sanjaiyan-dev** [comment]
Hi,
So according to my understanding, currently planned to `preload` the beforeInteractive scripts and `prefetch` the other scripts automatically without developers to specify. hope I understood it correctly 🙌

And,@housseindjirdeh, Oops I didn't if I can I will attach it ASAP (I am not that good at testing/benchmarking 😅)

**Reply by xd-hearst** [reply]
don't think there is correlation between preload and beforeinteractive, nor prefetch vs others. basically do not think those script loading priorities are being factored in much. 
As indicated by the implementation here
https://github.com/vercel/next.js/blob/canary/packages/next/client/script.tsx

**Reply by sanjaiyan-dev** [reply]
> don't think there is correlation between preload and beforeinteractive, nor prefetch vs others. basically do not think those script loading priorities are being factored in much. As indicated by the implementation here https://github.com/vercel/next.js/blob/canary/packages/next/client/script.tsx

Sorry, actually I thought to prioritize the beforeInteractive scripts mainly :)

**Comment by xd-hearst** [comment]
<strike>Any update regarding this? There is a distinction between `preload`, `prefetch` vs other means of script loading. This article has given some performance highlights: https://medium.com/reloading/preload-prefetch-and-priorities-in-chrome-776165961bbf

It is not clear how different next script strategies would/could accomplish the same thing</strike>

I have done some reading. Looks like Next js is doing the google-recommended thing. 

> This component is designed to "optimize the critical rendering path and ensure external scripts don't become a bottleneck to optimal page load."

https://www.patterns.dev/posts/loading-sequence/
